### PR TITLE
Enable mypy checking on V1 code

### PR DESCRIPTION
**Pull request by markmc**
As per #10959

Some of these definitely deserve closer scrutiny e.g.:

- ~~commit 7974f1d - was there a bug with the `hash_block_tokens()` call?~~
- ~~commit 226fd552089ea8f0aeb49c8c2a4cc62f2b32742a - feels we'll be fighting an uphill battle with mypy and this args/kwargs pattern~~
- ~~commit 8a6d88c42df22d9ef8d1deb444a9335117b91dc6 - `MultiprocExecutor.workers` and `UniprocExecutor.worker` can be set to `None`, so we have to litter the code with `is not None` assertions~~

### CODE REVIEW

File: vllm/v1/engine/core_client.py [hunk 1]
```diff
@@ -71,6 +74,9 @@ async def get_output_async(self) -> List[EngineCoreOutput]:
     async def add_request_async(self, request: EngineCoreRequest) -> None:
         raise NotImplementedError
 
+    async def profile_async(self, is_start=True) -> None:
```
[the remark this answers is not in the document]
Comment (markmc) [reply vllm/v1/engine/core_client.py]: Done across the board, thanks

File: vllm/v1/core/kv_cache_utils.py [hunk 2]
```diff
@@ -152,7 +152,7 @@ def get_all_free_blocks(self) -> List[KVCacheBlock]:
 
 
 def hash_block_tokens(parent_block_hash: Optional[int],
-                      curr_block_token_ids: Tuple[int]) -> BlockHashType:
+                      curr_block_token_ids: List[int]) -> BlockHashType:
```
[the remark this answers is not in the document]
Comment (markmc) [reply vllm/v1/core/kv_cache_utils.py:155]: It's actually the revers AFAICT - in hash_request_tokens() we were converting the token IDs from their natural list[int] form to tuple[int]

```
block_token_ids = tuple(token_ids[start:end])
```

I think just removing that conversion to tuple makes more sense?

[the remark this answers is not in the document]
Comment (markmc) [reply vllm/v1/core/kv_cache_utils.py]: ok, to avoid the conversion from ConstantList to list elsewhere, I've added:

```
TokenIdsType = Union[List[int], ConstantList[int]]
BlockHashType = Tuple[int, TokenIdsType]
```

Collection didn't work, because it needs to be indexable/sliceable:

```
    for start in range(0, len(token_ids), block_size):
        end = start + block_size
	block_token_ids = token_ids[start:end]
```

[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/v1/core/kv_cache_utils.py]: How about `Sequence[int]` then?

[the remark this answers is not in the document]
Comment (markmc) [reply vllm/v1/core/kv_cache_utils.py]: Good idea, done now :+1:

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -223,9 +223,10 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
 
         # Get the number of scheduled tokens for each request.
         # TODO: The Python loop can be slow. Optimize.
-        num_scheduled_tokens = []
+        num_scheduled_tokens = np.empty(0, dtype=np.int32)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:226]: Can we keep this a list but just rename it? I'd like to avoid using `append` for numpy arrays.

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -509,8 +512,14 @@ def execute_model(
             logprobs = None
         else:
             logprobs = sampler_output.logprobs.cpu()
+
+        assert all(
+            req_id is not None for req_id in
+            self.input_batch.req_ids[:num_reqs]), "req_ids contains None"
+        req_ids = cast(List[str], self.input_batch.req_ids[:num_reqs])
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:519]: QQ: why is this line needed?

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
Comment (markmc) [reply vllm/v1/worker/gpu_model_runner.py:259]: I'll revert this, I think #11111 removed the original error. Great catch :+1:

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (markmc) [reply vllm/v1/worker/gpu_model_runner.py:549]: This is the error:

```
vllm/v1/worker/gpu_model_runner.py:546: error: Argument "req_ids" to "ModelRunnerOutput" has incompatible type "list[Optional[str]]"; expected "list[str]"  [arg-type]
```

`input_batch.req_ids` can have `None` elements:

```
self.req_ids: List[Optional[str]] = [None] * max_num_reqs
```

but `ModelRunnerOutput.req_ids` can't:

```
class ModelRunnerOutput:

    # [num_reqs]                                                                                                                                                                                                                                                                  
    req_ids: List[str]
```

AFAIR `input_batch.num_reqs` tells us how many are non-None?

**Comment by markmc** [comment]
@DarkLight1337 @WoosukKwon note that I've rebased and fixed a bunch of new issues. Those new fixes are in the 5 latest commits

**Comment by markmc** [comment]
rebased again to fix conflicts

@comaniac would you mind reviewing commit 9190f342ffc4edfae851c4f0216ffc1231cde4ec ?

For reference, in #9972 you said:

> Enhance the block hash type from int to Tuple[int, Tuple[int]], which is (hash_value, (toke_ids,)). This guarantees no hash conflicts

Pasting the commit message:

The current block hash format is:

```
  (hash((parent_hash, *token_id_list)), token_id_list)
```

i.e. it is a tuple with two parts:

1. A hash of a tuple combining the parent hash with token ids
2. A tuple containing the same token ids

and this (hash, token_ids) tuple itself must be hashable in order to be a dictionary key.

This is suspicious - why include the same token IDs twice in the same hash?

And then this code:

```
def hash_request_tokens(block_size: int,
                        token_ids: List[int]) -> List[BlockHashType]:
    ...
    ret = []
    parent_block_hash = None
    for ...:
        tuple(token_ids[start:end])
        ...
        block_hash = hash_block_tokens(parent_block_hash, block_token_ids)
        parent_block_hash = block_hash
    ...
```

looks wrong because hash_block_tokens() expects an integer hash for the parent_block_hash argument:

```
def hash_block_tokens(parent_block_hash: Optional[int],
                      curr_block_token_ids: Tuple[int]) -> BlockHashType:
```

It appears like it would be perfectly sufficient to use the integer hash of the `(parent_hash + token_ids)` tuple.

If we do this, then the curr_block_token_ids argument no longer needs to be a tuple - but it does need to be a sliceable sequence, so let's make sure ConstantList satisfies that requirement.

**Comment by comaniac** [comment]
> This is suspicious - why include the same token IDs twice in the same hash?

This is to avoid potential hash collusion, because Python dict guarantees no collusion if the dict key is tuple.
But yeah the type is incorrect as you said. I'll submit a fix PR in minutes.

**Comment by markmc** [comment]
> > This is suspicious - why include the same token IDs twice in the same hash?
> 
> This is to avoid potential hash collision, because Python dict guarantees no collision if the dict key is tuple. But yeah the type is incorrect as you said. I'll submit a fix PR in minutes.

Thanks for the fix! Rebased again to resolve merge conflicts and fixed a few mypy errors with this prefix caching code.

**Comment by markmc** [comment]
Moving to draft while I debug the test failures

File: vllm/v1/executor/multiproc_executor.py
Comment (tlrmchlsmth) [inline vllm/v1/executor/multiproc_executor.py:1]: The asserts on the workers here are pretty nasty.

Setting the workers to `None` during `shutdown` was a bit of paranoia on my end to make sure everything was cleaned up.

I suggest changing shutdown to the following, and then we can ditch the asserts
```
    def shutdown(self):
        """Properly shut down the executor and its workers"""
        if getattr(self, 'shutting_down', False):
            self.shutting_down = True
            for w in self.workers:  #TODO: not sure if needed
                w.worker_response_mq = None
            self._ensure_worker_termination()

        self.rpc_broadcast_mq = None
```

Comment (markmc) [reply vllm/v1/executor/multiproc_executor.py:1]: Makes sense, thanks. Done now.

That's all 3 of my "dubious changes" sorted! :+1:

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:550]: Got it. Thanks!

File: vllm/v1/executor/multiproc_executor.py [hunk 5]
```diff
@@ -103,7 +104,7 @@ def collective_rpc(self,
                        method: str,
                        timeout: Optional[float] = None,
                        args: Tuple = (),
-                       kwargs: Optional[Dict] = None) -> []:
+                       kwargs: Optional[Dict] = None) -> Any:
```
Comment (tlrmchlsmth) [inline vllm/v1/executor/multiproc_executor.py:107]:
```suggestion
                       kwargs: Optional[Dict] = None) -> List[Any]:
```

File: vllm/v1/executor/uniproc_executor.py
Comment (tlrmchlsmth) [inline vllm/v1/executor/uniproc_executor.py:1]: Similar to `multiproc_executor`, we don't need to set worker to None in `shutdown`, which will let us remove the asserts

**Review by tlrmchlsmth — APPROVED** [review]
Thanks so much for this one!

File: vllm/v1/executor/multiproc_executor.py (hunk 5 above)
Comment (markmc) [reply vllm/v1/executor/multiproc_executor.py]: Thanks, good point

File: vllm/v1/executor/uniproc_executor.py
Comment (markmc) [reply vllm/v1/executor/uniproc_executor.py:1]: Doh. Of course. Done now.

After a rebase, another one of these crept in, but doesn't require asserts so I didn't touch it. See https://github.com/vllm-project/vllm/pull/11074/files#r1884597768

**Comment by WoosukKwon** [comment]
@markmc Finally merged! Thanks for your patience and great work! Really appreciate it.
